File: pie-mobile/adapter/host-gateway/src/relay-url.ts
export const DEFAULT_RELAY_ORIGIN =
  'https://relay.cookai.dev'

type RelayEnv = Record<string, string | undefined>

export function resolveRelayUrl(
  env: RelayEnv = process.env,
  explicitUrl?: string
): string {
  const relayUrl = explicitUrl?.trim() || env.PIE_RELAY_URL?.trim() || DEFAULT_RELAY_ORIGIN
  const parsed = new URL(relayUrl)
  if (!['http:', 'https:', 'ws:', 'wss:'].includes(parsed.protocol)) {
    throw new Error('Relay URL must use http(s) or ws(s)')
  }
  parsed.pathname = '/'
  parsed.search = ''
  parsed.hash = ''
  return parsed.origin
}

File: executor-manager/cmd/manager/relay_url.go
package main

import (
	"errors"
	"net/url"
	"strings"
)

const defaultRelayOrigin = "https://relay.cookai.dev"

func resolveRelayAgentURL(raw string) (string, error) {
	selected := strings.TrimSpace(raw)
	if selected == "" {
		selected = defaultRelayOrigin
	}
	parsed, err := url.Parse(selected)
	if err != nil || parsed.Host == "" {
		return "", errors.New("PIE_RELAY_URL is invalid")
	}
	switch parsed.Scheme {
	case "http":
		parsed.Scheme = "ws"
	case "https":
		parsed.Scheme = "wss"
	case "ws", "wss":
	default:
		return "", errors.New("PIE_RELAY_URL must use http(s) or ws(s)")
	}
	path := strings.TrimRight(parsed.Path, "/")
	if path == "" {
		path = "/ws/agent"
	} else if !strings.HasSuffix(path, "/ws/agent") {
		path += "/ws/agent"
	}
	parsed.Path, parsed.RawPath = path, ""
	parsed.RawQuery, parsed.Fragment = "", ""
	return parsed.String(), nil
}
